### PR TITLE
feature/file-schema-auto-detection

### PR DESCRIPTION
**Pull request by ailegion**
* added: SchemaAutodetection support for delimited formats
* added: Schema Auto-detection support for Avro format
* added: Schema Auto-detection support for Parquet format
* added: Get Schema button to File source widget

### CODE REVIEW

File: format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedConfig.java [hunk 1]
```diff
@@ -49,4 +67,101 @@
...
+    final Path file = getFilePathForSchemaGeneration(path,
+        format.equals("delimited") ? null : format, conf);
+    final FileSystem fileSystem = FileSystem.get(file.toUri(), conf);
+    FSDataInputStream input = fileSystem.open(file);
+    BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(input));
+    String line = bufferedReader.readLine();
+    if (line == null) {
+      return null;
+    }
+    bufferedReader.close();
```
Comment (nitinmotgi) [inline format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedConfig.java]: Close should be final

File: format-avro/src/main/java/io/cdap/plugin/format/avro/input/AvroInputFormatProvider.java [hunk 2]
```diff
@@ -56,4 +76,50 @@ protected void addFormatProperties(Map<String, String> properties) {
...
+          .openReader(seekableInput, dataReader);
+      GenericRecord firstRecord;
+      if (!dataFileReader.hasNext()) {
+        return null;
+      }
+      firstRecord = dataFileReader.next();
+      return new AvroToStructuredTransformer().convertSchema(firstRecord.getSchema());
+    } catch (IOException e) {
+      context.getFailureCollector().addFailure("Schema parse error", e.getMessage());
+    }
```
Comment (nitinmotgi) [inline format-avro/src/main/java/io/cdap/plugin/format/avro/input/AvroInputFormatProvider.java:150]: Close readers

File: format-parquet/src/main/java/io/cdap/plugin/format/parquet/input/ParquetInputFormatProvider.java [hunk 3]
```diff
@@ -56,4 +71,37 @@ protected void addFormatProperties(Map<String, String> properties) {
...
+  public Schema getSchema(FormatContext context) {
+    if (conf.containsMacro(PathTrackingConfig.NAME_SCHEMA) || !Strings.isNullOrEmpty(conf.schema)) {
+      return super.getSchema(context);
+    }
+    String filePath = conf.getProperties().getProperties().getOrDefault("path", null);
+
+    try {
+      Job job = JobUtils.createInstance();
+      Configuration hconf = job.getConfiguration();
+      final Path file = conf.getFilePathForSchemaGeneration(filePath, "parquet", hconf);
```
Comment (nitinmotgi) [inline format-parquet/src/main/java/io/cdap/plugin/format/parquet/input/ParquetInputFormatProvider.java:105]: Reader needs closing ? If so then it should be in final block

**Review by nitinmotgi — COMMENTED** [review]
Few comments to look at.

**Comment by CuriousVini** [comment]
Also could you please open PR against develop branch first. Usually we follow the practice of opening first PR against develop and cherry-picking to release branches if required

**Comment by ailegion** [comment]
> Also could you please open PR against develop branch first. Usually we follow the practice of opening first PR against develop and cherry-picking to release branches if required

Created PR against develop branch: https://github.com/cdapio/hydrator-plugins/pull/1203

**Pull request closed without merging**